firestore/query.go

```diff
@@ -579,57 +579,18 @@ func trunc32(i int) int32 {
 	return int32(i)
 }
 
-// GetAll returns an array of query's resulting documents.
-func (q Query) GetAll(ctx context.Context) ([]*DocumentSnapshot, error) {
-	limitedToLast := q.limitToLast
-
-	if q.limitToLast {
-		// Flip order statements before posting a request.
-		for i := range q.orders {
-			if q.orders[i].dir == Asc {
-				q.orders[i].dir = Desc
-			} else {
-				q.orders[i].dir = Asc
-			}
-		}
-		// Swap cursors.
-		q.startVals, q.endVals = q.endVals, q.startVals
-		q.startDoc, q.endDoc = q.endDoc, q.startDoc
-		q.startBefore, q.endBefore = q.endBefore, q.startBefore
-
-		q.limitToLast = false
-	}
-	docs, err := q.Documents(ctx).GetAll()
-	if err != nil {
-		return nil, err
-	}
-	if limitedToLast {
-		// Flip docs order before return.
-		for i, j := 0, len(docs)-1; i < j; {
-			docs[i], docs[j] = docs[j], docs[i]
-			i++
-			j--
-		}
-	}
-	return docs, nil
-}
-
 // Documents returns an iterator over the query's resulting documents.
 func (q Query) Documents(ctx context.Context) *DocumentIterator {
-	if q.limitToLast {
-		return &DocumentIterator{
-			err: errors.New("firestore: queries that include limitToLast constraints cannot be streamed. Use Query.GetAll() instead"),
-		}
-	}
 	return &DocumentIterator{
-		iter: newQueryDocumentIterator(withResourceHeader(ctx, q.c.path()), &q, nil),
+		iter: newQueryDocumentIterator(withResourceHeader(ctx, q.c.path()), &q, nil), q: q,
 	}
 }
 
 // DocumentIterator is an iterator over documents returned by a query.
 type DocumentIterator struct {
 	iter docIterator
 	err  error
+	q    Query
```

Comment:
As far as the Query being included here I don't expect there to be a issue with memory since the query itself should already be built up in memory by the time this is created. The potential memory pressure comes from the need to buffer the results when `limitToLast` is `true`, but that is expected given the design of the feature.

Reply:
Actually I see that `queryDocumentIterator` uses a `*Query` instead, so we should probably do that for consistency.

Reply:
@IlyaFaer could you switch this to use a `*Query` instead? Otherwise this is looking pretty much good to go to me.

Reply:
@tritone, sure. I've already written it and stashed for some past reasons. Pushing